src/resolver/resolver.zig

```diff
@@ -1952,8 +1961,32 @@ pub const Resolver = struct {
                     esm_resolution.status = .ModuleNotFound;
                     return null;
                 };
-                const entry_query = entries.get(std.fs.path.basename(abs_esm_path)) orelse {
+                const base = std.fs.path.basename(abs_esm_path);
+                const entry_query = entries.get(base) orelse {
+                    const ends_with_star = esm_resolution.status == .ExactEndsWithStar;
                     esm_resolution.status = .ModuleNotFound;
+
+                    // Try to have a friendly error message if people forget the extension
+                    if (ends_with_star) {
+                        const extension_order = if (kind == .at or kind == .at_conditional)
+                            r.extension_order
+                        else
+                            r.opts.extension_order;
+
+                        std.mem.copy(u8, &load_as_file_buf, base);
+                        for (extension_order) |ext| {
+                            var file_name = load_as_file_buf[0 .. base.len + ext.len];
+                            std.mem.copy(u8, file_name[base.len..], ext);
+                            if (entries.get(file_name) != null) {
+                                if (r.debug_logs) |*debug| {
+                                    debug.addNoteFmt("The import {s} is missing the extension {s}", .{ ResolvePath.join(.{ package_json.name, esm_subpath_buf }, .auto), ext });
```

Comment:
`esm_subpath_buf` is the entire buffer preallocated buffer, but it needs to be a small slice of that buffer because most of the path within `esm_subpath_buf` is not in use. This will cause printing undefined memory into the logs.

Reply:
Ok, that makes sense, I should be using `package_json.subpath` instead, right?

Reply:
yeah

Reply:
`subpath` is actually a property on `ESModule.Package` rather than `PackageJSON`, I misread it yesterday. I updated the call-sites to explicitly pass the `esm.subpath` string, but do let me know if you see a more elegant way to do the same thing.